### PR TITLE
Pairwise aligner

### PR DESCRIPTION
**Pull request by mdehoon**
This pull request provides a new implementation of pairwise aligners in Bio.Align as a possible replacement for the pairwise alignment code in Bio.pairwise2. The main advantages are that the code is more Pythonic, and more general than Bio.pairwise2. Detailed information on how to use the pairwise alignment code can be found in the chapter_align.tex file under Doc/Tutorial/. The pairwise alignments are represented by PairwiseAlignment objects; it may be better to integrate this with the existing MultipleSeqAlignment class.

I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

I have read the ``CONTRIBUTING.rst`` file and understand that AppVeyor and
TravisCI will be used to confirm the Biopython unit tests and ``flake8`` style
checks pass with these changes.

I am happy be thanked by name in the ``NEWS.rst`` and ``CONTRIB.rst`` files,
and have added myself to those files as part of this pull request. (*This
acknowledgement is optional. Note we list the names sorted alphabetically.*)

### CODE REVIEW

File: Bio/Align/__init__.py [hunk 1]
```diff
@@ -923,6 +925,405 @@ def sort(self, key=None, reverse=False):
...
+        n2 -= end2
+        n = max(n1, n2)
+        aligned_seq1 += seq1[end1:] + ' ' * (n - n1)
+        aligned_seq2 += seq2[end2:] + ' ' * (n - n2)
+        pattern += ' ' * n
+        text = """\
+%s
+%s
+%s
+""" % (aligned_seq1, pattern, aligned_seq2)
```
Comment (peterjc) [inline Bio/Align/__init__.py]: Would this be clearer as:

```python
return "%s\n%s\n%s\n" % (aligned_seq1, pattern, aligned_seq2)
```

File: Bio/Align/__init__.py [hunk 2]
```diff
@@ -923,6 +925,405 @@ def sort(self, key=None, reverse=False):
...
+        return self
+
+    def __next__(self):
+        path = next(self.paths)
+        self.index += 1
+        alignment = PairwiseAlignment(self.seqA, self.seqB, path, self.score)
+        self.alignment = alignment
+        return alignment
+
+    next = __next__  # Python 2
```
Comment (peterjc) [inline Bio/Align/__init__.py]: Do we need to define the ``.next()`` method at all? Python 2.7 supports the built-in function ``next(...)`` which would work on both.

Or can we make this conditional on Python 2, like:
https://github.com/biopython/biopython/blob/biopython-171/Bio/File.py#L156

**Comment by peterjc** [comment]
What do you think @MarkusPiotrowski? Does this provide a nicer API? Are there evil corner test cases we should port over to this module?

**Comment by peterjc** [comment]
@mdehoon are you familiar with ``git rebase origin/master`` as one way to deal with the merge conflict in ``setup.py``?

**Comment by mdehoon** [comment]
@peterjc  Not sure what rebasing did here

**Comment by peterjc** [comment]
Hmm, the rebase didn't look like what I'd have expected, but I think I know why - you probably rebased to an older copy of the master? Retry:

```
# Assuming on branch pairwise_aligner and
# remotes are called origin and mdehoon
git fetch origin
git rebase origin/master
git push mdehoon pairwise_aligner --force
```

**Comment by mdehoon** [comment]
@peterjc  I'm getting new merge errors all over the place. Perhaps it's easier to start from a fresh pull request.

**Comment by peterjc** [comment]
If that is simplest for you, sure.

I tried rebasing this myself, the ``setup.py`` file was annoying, but once it got into the alignment doctest changes I thought it best you do this.

**Comment by mdehoon** [comment]
Closing this pull request; see pull request #1655  for a clean pull request for the pairwise aligner code.

**Pull request closed without merging**